src/javascript/expression.rs

```diff
@@ -210,6 +239,233 @@ impl<'module> Generator<'module> {
+            }
+            Pattern::Tuple { elems, .. } => {
+                // We don't check the length, because type system means it's a tuple
+                for (index, pattern) in elems.into_iter().enumerate() {
+                    let mut path = path.clone();
+                    path.push(Document::String(format!("{}", index)));
+                    self.traverse_pattern(pattern, tmp_var.clone(), &mut path, checks, assignments)?
+                }
+                Ok(())
+            }
+            Pattern::Constructor {
+                constructor: PatternConstructor::Record { name, .. },
+                ..
+            } if name == "True" => {
+                checks.push(equality(tmp_var, &path, "true".to_doc()));
+                Ok(())
+            }
+            Pattern::Constructor {
+                constructor: PatternConstructor::Record { name, .. },
+                ..
+            } if name == "False" => {
+                checks.push(equality(tmp_var, &path, "false".to_doc()));
+                Ok(())
+            }
+            Pattern::Constructor {
+                constructor: PatternConstructor::Record { name, .. },
+                ..
+            } if name == "Nil" => {
+                checks.push(equality(tmp_var, &path, "undefined".to_doc()));
+                Ok(())
+            }
+            Pattern::Constructor {
+                constructor: PatternConstructor::Record { name, field_map },
+                arguments,
+                ..
+            } => {
+                let mut type_path = path.clone();
+                type_path.push(string("type"));
+                checks.push(equality(tmp_var.clone(), &type_path, string(name)));
+
+                let _ = arguments
+                    .iter()
+                    .enumerate()
+                    .map(|x| {
+                        let (index, arg) = x;
+                        let mut path = path.clone();
+                        match field_map {
+                            None => path.push(Document::String(format!("{}", index))),
+                            Some(FieldMap { fields, .. }) => {
+                                let label = fields.iter().find_map(|(key, &val)| {
+                                    if val == index {
+                                        Some(key)
+                                    } else {
+                                        None
+                                    }
+                                });
+                                path.push(string(label.unwrap()))
+                            }
+                        }
+                        self.traverse_pattern(
+                            &arg.value,
+                            tmp_var.clone(),
+                            &mut path,
+                            checks,
+                            assignments,
+                        )
+                    })
+                    .collect::<Result<Vec<()>, Error>>()?;
```

Comment:
Can we remove the vector building here? It is discarded immediately. A for loop will work as alternative.

Reply:
Done